File: app/securesocial/provider/OAuth1Provider.java
/**
* Copyright 2011 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
*/
package securesocial.provider;

import play.Play;
import play.cache.Cache;
import play.libs.OAuth;
import play.mvc.Scope;
import play.mvc.results.Redirect;

import java.util.Map;

/**
 * A provider that handles the OAuth1 authentication flow
 */
public abstract class OAuth1Provider extends IdentityProvider {

    private static final String DENIED = "denied";
    private static final String SECURESOCIAL = "securesocial.";
    private OAuth.ServiceInfo serviceInfo;
    private static final String REQUEST_TOKEN_URL = "requestTokenURL";
    private static final String ACCESS_TOKEN_URL = "accessTokenURL";
    private static final String AUTHORIZATION_URL = "authorizationURL";
    private static final String CONSUMER_KEY = "consumerKey";
    private static final String CONSUMER_SECRET = "consumerSecret";


    protected OAuth1Provider(ProviderType type)
    {
        super(type, AuthenticationMethod.OAUTH1);
        serviceInfo = createServiceInfo(getPropertiesKey(type));
    }

    /**
     * Gets the service info required to invoke the APIs for this provider.
     *
     * @return A play.libs.OAuth.ServiceInfo object
     */
    public OAuth.ServiceInfo getServiceInfo() {
        return serviceInfo;
    }

    /**
     * Creates a ServiceInfo using the properties in the application.conf
     *
     * @param key The key for this provider
     * @return A OAuth.ServiceInfo object
     */
    public static OAuth.ServiceInfo createServiceInfo(String key) {
        return new OAuth.ServiceInfo(
                Play.configuration.getProperty(key + REQUEST_TOKEN_URL),
                Play.configuration.getProperty(key + ACCESS_TOKEN_URL),
                Play.configuration.getProperty(key + AUTHORIZATION_URL),
                Play.configuration.getProperty(key + CONSUMER_KEY),
                Play.configuration.getProperty(key + CONSUMER_SECRET)
        );

    }

    @Override
    public SocialUser doAuth(Map<String, Object> authContext) throws AccessDeniedException {
        Scope.Params params = Scope.Params.current();

        if ( params.get(DENIED) != null ) {
            throw new AccessDeniedException();
        }

        final String key = new StringBuilder(SECURESOCIAL).append(Scope.Session.current().getId()).toString();
        OAuth service = OAuth.service(serviceInfo);

        if ( !OAuth.isVerifierResponse() ) {
            // first step on the authentication process
            OAuth.Response response = service.retrieveRequestToken();
            if ( response.error != null ) {
                // there was an error retrieving the access token
                throw new AuthenticationException(response.error);
            }
            SocialUser user = createUser();
            user.token = response.token;
            user.secret = response.secret;
            Cache.add(key, user);
            throw new Redirect( service.redirectUrl(response.token), false);
        }

        // the OAuth provider is redirecting back to us
        SocialUser user = (SocialUser) Cache.get(key);
        if ( user == null ) {
            throw new AuthenticationException();
        }
        Cache.delete(key);
        OAuth.Response response = service.retrieveAccessToken(user.token, user.secret);
        if ( response.error != null ) {
            throw new AuthenticationException(response.error);
        }

        // all was ok, replace the tokens
        user.token = response.token;
        user.secret = response.secret;
        user.serviceInfo = serviceInfo;
        return user;
    }
}

File: module-code/app/securesocial/core/OAuth1Provider.scala
/**
 * Copyright 2012-2014 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.core

import _root_.java.util.UUID

import com.typesafe.config.ConfigObject
import io.methvin.play.autoconfig.AutoConfig
import play.api.libs.json.JsValue
import play.api.libs.oauth.{ ConsumerKey, OAuth, RequestToken, ServiceInfo, _ }
import play.api.mvc.Results.Redirect
import play.api.mvc.{ AnyContent, Request }
import play.api.{ ConfigLoader, Configuration }
import play.shaded.oauth.oauth.signpost.exception.OAuthException
import securesocial.core.services.{ CacheService, HttpService, RoutesService }

import scala.concurrent.{ ExecutionContext, Future }

/**
 * A trait that allows mocking the OAuth 1 client
 */
trait OAuth1Client {

  def retrieveRequestToken(callbackURL: String): Future[RequestToken]

  def retrieveOAuth1Info(token: RequestToken, verifier: String): Future[OAuth1Info]

  def redirectUrl(token: String): String

  def retrieveProfile(url: String, info: OAuth1Info): Future[JsValue]

  implicit def executionContext: ExecutionContext
}

object OAuth1Client {
  /**
   * A default implementation based on the Play client
   * @param serviceInfo
   */
  class Default(val serviceInfo: ServiceInfo, val httpService: HttpService)(implicit val executionContext: ExecutionContext) extends OAuth1Client {
    private[core] val client = OAuth(serviceInfo, use10a = true)
    override def redirectUrl(token: String): String = client.redirectUrl(token)

    private def withFuture(call: => Either[OAuthException, RequestToken]): Future[RequestToken] = Future {
      call match {
        case Left(error) => throw error
        case Right(token) => token
      }
    }

    override def retrieveOAuth1Info(token: RequestToken, verifier: String) = withFuture {
      client.retrieveAccessToken(token, verifier)
    }.map(accessToken => OAuth1Info(accessToken.token, accessToken.secret))

    override def retrieveRequestToken(callbackURL: String) = withFuture {
      client.retrieveRequestToken(callbackURL)
    }

    override def retrieveProfile(url: String, info: OAuth1Info): Future[JsValue] =
      httpService.url(url).sign(OAuthCalculator(serviceInfo.key, RequestToken(info.token, info.secret))).get().map(_.json)
  }
}

case class OAuth1Settings(
  requestTokenUrl: String,
  accessTokenUrl: String,
  authorizationUrl: String,
  consumerKey: String,
  consumerSecret: String)
object OAuth1Settings {
  implicit val configLoader: ConfigLoader[OAuth1Settings] = AutoConfig.loader
  def forProvider(configuration: Configuration, id: String): OAuth1Settings = {
    val path = s"securesocial.$id"
    val defaultPath = "securesocial.oauth1Settings"
    (Configuration(path -> configuration.get[ConfigObject](defaultPath)) ++ configuration).get[OAuth1Settings](path)
  }
}

object ServiceInfoHelper {
  /**
   * A helper method to create a service info from the properties file
   * @param id
   * @return
   */
  def forProvider(configuration: Configuration, id: String): ServiceInfo = {
    val config = configuration.get[OAuth1Settings](s"securesocial.$id")
    import config._
    ServiceInfo(requestTokenUrl, accessTokenUrl, authorizationUrl, ConsumerKey(consumerKey, consumerSecret))
  }
}

/**
 * Base class for all OAuth1 providers
 */
abstract class OAuth1Provider(
  routesService: RoutesService,
  cacheService: CacheService,
  val client: OAuth1Client)
  extends IdentityProvider {

  protected implicit val executionContext = client.executionContext
  protected val logger = play.api.Logger(this.getClass.getName)

  def authMethod = AuthenticationMethod.OAuth1

  def authenticate()(implicit request: Request[AnyContent]): Future[AuthenticationResult] = {
    if (request.queryString.get("denied").isDefined) {
      // the user did not grant access to the account
      Future.successful(AuthenticationResult.AccessDenied())
    } else {
      val verifier = request.queryString.get("oauth_verifier").map(_.head)
      if (verifier.isEmpty) {
        // this is the 1st step in the auth flow. We need to get the request tokens
        val callbackUrl = routesService.authenticationUrl(id)
        logger.debug("[securesocial] callback url = " + callbackUrl)
        client.retrieveRequestToken(callbackUrl).flatMap { accessToken =>
          val cacheKey = UUID.randomUUID().toString
          val redirect = Redirect(client.redirectUrl(accessToken.token))
            .withSession(request.session + (OAuth1Provider.CacheKey -> cacheKey))
          // set the cache key timeoutfor 5 minutes, plenty of time to log in
          cacheService
            .set(cacheKey, accessToken, 300)
            .map(_ => AuthenticationResult.NavigationFlow(redirect))
        } recover {
          case e =>
            logger.error("[securesocial] error retrieving request token", e)
            throw AuthenticationException()
        }
      } else {
        // 2nd step in the oauth flow
        val cacheKey = request.session.get(OAuth1Provider.CacheKey).getOrElse {
          logger.error("[securesocial] missing cache key in session during OAuth1 flow")
          throw AuthenticationException()
        }
        for (
          requestToken <- cacheService.getAs[RequestToken](cacheKey).recover {
            case e =>
              logger.error("[securesocial] error retrieving entry from cache", e)
              throw AuthenticationException()
          };
          accessToken <- client.retrieveOAuth1Info(
            RequestToken(requestToken.get.token, requestToken.get.secret), verifier.get).recover {
              case e =>
                logger.error("[securesocial] error retrieving access token", e)
                throw AuthenticationException()
            };
          result <- fillProfile(accessToken)
        ) yield {
          AuthenticationResult.Authenticated(result)
        }
      }
    }
  }

  def fillProfile(info: OAuth1Info): Future[BasicProfile]
}

object OAuth1Provider {
  val CacheKey = "cacheKey"
  val RequestTokenUrl = "requestTokenUrl"
  val AccessTokenUrl = "accessTokenUrl"
  val AuthorizationUrl = "authorizationUrl"
  val ConsumerKey = "consumerKey"
  val ConsumerSecret = "consumerSecret"
}
